src/Middleware/WebSockets/test/UnitTests/WebSocketMiddlewareTests.cs

```diff
@@ -795,6 +795,57 @@ public async Task AcceptingWebSocketRequestDisablesTimeout()
         }
     }
 
+    [Fact]
+    public async Task PingTimeoutCancelsReceiveAsync()
+    {
+        var tcs = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);
+        await using (var server = KestrelWebSocketHelpers.CreateServer(LoggerFactory, out var port, async context =>
+        {
+            try
+            {
+                Assert.True(context.WebSockets.IsWebSocketRequest);
+                var webSocket = await context.WebSockets.AcceptWebSocketAsync();
+                await webSocket.ReceiveAsync(new byte[1], cancellationToken: default);
+            }
+            catch (Exception ex)
+            {
+                tcs.TrySetException(ex);
+            }
+            finally
+            {
+                tcs.TrySetResult();
+            }
+        },
+        o =>
+        {
+            o.KeepAliveInterval = TimeSpan.FromMilliseconds(1);
+            o.KeepAliveTimeout = TimeSpan.FromMilliseconds(1);
+        }))
+        {
+            using (var client = new HttpClient())
+            {
+                var uri = new UriBuilder(new Uri($"ws://127.0.0.1:{port}/"));
+                uri.Scheme = "http";
+
+                // Craft a valid WebSocket Upgrade request
+                using (var request = new HttpRequestMessage(HttpMethod.Get, uri.ToString()))
+                {
+                    request.Headers.Connection.Clear();
+                    request.Headers.Connection.Add("Upgrade");
+                    request.Headers.Upgrade.Add(new System.Net.Http.Headers.ProductHeaderValue("websocket"));
+                    request.Headers.Add(HeaderNames.SecWebSocketVersion, "13");
+                    // SecWebSocketKey required to be 16 bytes
+                    request.Headers.Add(HeaderNames.SecWebSocketKey, Convert.ToBase64String(new byte[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16 }, Base64FormattingOptions.None));
+
+                    var response = await client.SendAsync(request, HttpCompletionOption.ResponseHeadersRead);
+
+                    var ex = await Assert.ThrowsAnyAsync<Exception>(() => tcs.Task);
+                    Assert.True(ex is ConnectionAbortedException or WebSocketException, ex.GetType().FullName);
```

Comment:
I figured you'd always get a `ConnectionAbortedException` if the read is canceled by `HttpContext.Abort()`. Is it ever really a `WebSocketException`?

Reply:
Yes, if the WebSocket.ReceiveAsync call occurs after the timeout. It's a race, we could reduce the chance by changing KeepAliveTimeout to a larger value, but then the test would take longer 😄 
And would still technically be a flaky test.